### PR TITLE
[custom] Add custom filter to use llama2c

### PR DESCRIPTION
**Pull request by niley7464**
This PR adds custom filter to use llama2c.

### How to get model and tokenizer
https://github.com/karpathy/llama2.c/tree/master?tab=readme-ov-file#custom-tokenizers
I tested with vocab size 4096.

### How to run
```
$ls
libnnstreamer_customfilter_llama2.so model.bin input.txt tokenizer.bin

$cat input.txt
There was a little

$ gst-launch-1.0 filesrc location=input.txt ! application/octet-stream ! tensor_converter ! other/tensors,format=flexible ! tensor_filter framework=custom model=libnnstreamer_customfilter_llama2.so ! filesink location=output.txt

$cat output.txt 
There was a little 3 year old girl named Lucy. She was very excited because today was her birthday. She was so happy and couldn't wait to open her presents.
Lucy's mom said, "Let's go to the store and get a present for you."
Lucy was so excited and quickly ran to the store. She saw so many presents and couldn't decide which one to get.
She asked her mom, "Which one should I get?"
Her mom said, "Let's get the one you want."
Lucy was so happy and said, "I want the one!"
Her mom smiled and said, "I'm sure it's a great present!"
Lucy was so excited and couldn't wait to open the present. She opened it and found a beautiful doll. She hugged it and said, "Thank you, Mommy!"
Her mom smiled and said, "You're welcome, Lucy. I'm glad you like it."
```

See also : https://github.com/nnsuite/llama2.c

https://github.com/nnsuite/llama2.c/pull/4 is needed.

**Self evaluation:**
1. Build test: [*]Passed [ ]Failed [ ]Skipped
2. Run test: [*]Passed [ ]Failed [ ]Skipped

### CODE REVIEW

**Comment by niley7464** [comment]
Q)
tensor converter needs the exact input-dim like `3584:1:1:1`
I found the documentation about tensor_converter and it says
```
The size of a text frame should be configured by developer with the property input-dim.
Because the dimension of tensor is the key metadata of a tensor stream pipeline,
we need to fix the value before actually looking at the actual stream data.
```

Does anyone know how to use dimension flexbile when using text?

**Comment by anyj0527** [comment]
> Q) tensor converter needs the exact input-dim like `3584:1:1:1` I found the documentation about tensor_converter and it says
> 
> ```
> The size of a text frame should be configured by developer with the property input-dim.
> Because the dimension of tensor is the key metadata of a tensor stream pipeline,
> we need to fix the value before actually looking at the actual stream data.
> ```
> 
> Does anyone know how to use dimension flexbile when using text?

May use `application/octet-stream` and `other/tensors,format=flexible` with filter supporting flexible input tensor?
`gst-launch-1.0 filesrc location=input.txt ! application/octet-stream ! tensor_converter ! other/tensors,format=flexible ! tensor_filter framework=custom model=libsomeflexiblefilter.so ! filesink location=output.txt`

**Comment by niley7464** [comment]
> 
> May use `application/octet-stream` and `other/tensors,format=flexible` with filter supporting flexible input tensor? `gst-launch-1.0 filesrc location=input.txt ! application/octet-stream ! tensor_converter ! other/tensors,format=flexible ! tensor_filter framework=custom model=libsomeflexiblefilter.so ! filesink location=output.txt`

Your suggestion works well
Great thanks 👍

**Review by wooksong — APPROVED** [review]
LGTM!

File: tests/nnstreamer_example/custom_example_llama2/nnstreamer_customfilter_llama2.c [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+ */
+static int
+pt_invoke (void *private_data, const GstTensorFilterProperties *prop,
+    const GstTensorMemory *input, GstTensorMemory *output)
+{
+  pt_data *data = private_data;
+  char *result;
+  char *prompt = (char *) input[0].data;
+  int len;
+  UNUSED (prop);
```
Comment (songgot) [inline tests/nnstreamer_example/custom_example_llama2/nnstreamer_customfilter_llama2.c:141]: Good tests :+1: 
How do you think about checking if variables are NULL or not?

File: tests/nnstreamer_example/custom_example_llama2/nnstreamer_customfilter_llama2.c [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+/**
+ * @brief exit callback of tensor_filter custom
+ */
+static void
+pt_exit (void *private_data, const GstTensorFilterProperties *prop)
+{
+  pt_data *data = private_data;
+  UNUSED (prop);
+
```
Comment (songgot) [inline tests/nnstreamer_example/custom_example_llama2/nnstreamer_customfilter_llama2.c:90]: How do you think about checking if ```data``` is NULL or not before use?

File: tests/nnstreamer_example/custom_example_llama2/nnstreamer_customfilter_llama2.c [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+/**
+ * @brief setInputDimension callback of tensor_filter custom
+ */
+static int
+get_inputDim (void *private_data, const GstTensorFilterProperties *prop, GstTensorsInfo *info)
+{
+  UNUSED (private_data);
+  UNUSED (prop);
+
+  info->format = _NNS_TENSOR_FORMAT_SPARSE;
```
Comment (songgot) [inline tests/nnstreamer_example/custom_example_llama2/nnstreamer_customfilter_llama2.c:111]: here, info is necessary to check whether it is NULL. right ? :smile:

File: tests/nnstreamer_example/custom_example_llama2/nnstreamer_customfilter_llama2.c (hunk 1 above)
Comment (niley7464) [reply tests/nnstreamer_example/custom_example_llama2/nnstreamer_customfilter_llama2.c]: Added null check
Thank you :)

Comment (songgot) [reply tests/nnstreamer_example/custom_example_llama2/nnstreamer_customfilter_llama2.c]: Thank you very much! :smile:

**Review by songgot — APPROVED** [review]
Greate PR :100:

**Pull request merged**